File: native/src/android/jni_app/objects/app/comic_book_page.rs
/*
 * This file is part of Seeneva Android Reader
 * Copyright (C) 2021 Sergei Solodovnikov
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

use std::convert::TryInto;

use jni::errors::Result as JniResult;
use jni::objects::{JObject, JValue};
use jni::JNIEnv;
use once_cell::sync::Lazy;

use crate::comics::content::ComicPage;

use super::{comic_book_page_object, constants, JniClassConstructorCache};

static CONSTRUCTOR: Lazy<JniClassConstructorCache<&str, String>> = Lazy::new(|| {
    (
        constants::COMIC_BOOK_PAGE_TYPE,
        format!(
            "(\
    JLjava/lang/String;\
    I\
    I\
    [L{};\
    )V",
            constants::COMIC_BOOK_PAGE_OBJECT_TYPE
        ),
    )
        .into()
});

pub type ComicPageMetadata<'a> = JObject<'a>;
pub type ComicPageMetadataArray<'a> = JObject<'a>;

///Map data from Rust implementation to it Java variant
pub fn new<'b>(env: &'b JNIEnv, comic_page: &ComicPage) -> JniResult<ComicPageMetadata<'b>> {
    CONSTRUCTOR.init(env).and_then(|constructor| {
        constructor.create(&[
            JValue::Long(
                comic_page
                    .pos
                    .try_into()
                    .expect("Can't convert page position to Long"),
            ),
            env.new_string(&comic_page.name)?.into(),
            JValue::Int(
                comic_page
                    .width
                    .try_into()
                    .expect("Can't convert page width to Int"),
            ),
            JValue::Int(
                comic_page
                    .height
                    .try_into()
                    .expect("Can't convert page height to Int"),
            ),
            env.auto_local(comic_book_page_object::new_array(env, &comic_page.b_boxes)?)
                .as_obj()
                .into(),
        ])
    })
}

///Create array from provided slice [`comic_pages`]
pub fn new_array<'b>(
    env: &'b JNIEnv,
    comic_pages: &[ComicPage],
) -> JniResult<ComicPageMetadataArray<'b>> {
    env.with_local_frame(comic_pages.len() as _, || {
        let array = env.new_object_array(
            comic_pages.len() as _,
            CONSTRUCTOR.init(env)?.class_obj(),
            JObject::null(),
        )?;

        for (pos, page_metadata) in comic_pages.iter().enumerate() {
            let page_metadata = new(env, page_metadata)?;
            env.set_object_array_element(array, pos as _, page_metadata)?;
        }

        Ok(JObject::from(array))
    })
}

File: data/src/main/kotlin/app/seeneva/reader/data/entity/ComicBookPage.kt
/*
 *  This file is part of Seeneva Android Reader
 *  Copyright (C) 2021-2023 Sergei Solodovnikov
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package app.seeneva.reader.data.entity

import androidx.annotation.Keep
import androidx.room.*

/**
 * Metadata of one single page in the comic book
 * @param position page position in the comic container
 * @param name page file displayName
 * @param width comic page width
 * @param height comic page height
 */
@Entity(
    tableName = ComicBookPage.TABLE_NAME,
    foreignKeys = [ForeignKey(
        entity = ComicBook::class,
        parentColumns = [ComicBook.COLUMN_ID],
        childColumns = [ComicBookPage.COLUMN_BOOK_ID],
        onDelete = ForeignKey.CASCADE
    )],
    indices = [Index(
        name = ComicBookPage.INDEX_BOOK_ID_POSITION,
        value = [ComicBookPage.COLUMN_BOOK_ID, ComicBookPage.COLUMN_POSITION],
        unique = true
    )]
)
data class ComicBookPage @Ignore constructor(
    @PrimaryKey(autoGenerate = true)
    @ColumnInfo(name = COLUMN_ID)
    val id: Long,
    @ColumnInfo(name = COLUMN_BOOK_ID)
    val bookId: Long,
    @ColumnInfo(name = COLUMN_POSITION)
    val position: Long,
    @ColumnInfo(name = COLUMN_NAME)
    val name: String,
    @ColumnInfo(name = COLUMN_WIDTH)
    val width: Int,
    @ColumnInfo(name = COLUMN_HEIGHT)
    val height: Int,
    @Ignore
    val objects: List<ComicPageObject>
) {
    // for some reason @kotlin.jvm.JvmOverloads doesn't work anymore
    // https://issuetracker.google.com/issues/70762008
    /**
     * Used by room
     */
    internal constructor(
        id: Long,
        bookId: Long,
        position: Long,
        name: String,
        width: Int,
        height: Int
    ) : this(id, bookId, position, name, width, height, emptyList())

    /**
     * Called from a native side
     */
    @Suppress("unused")
    @Keep
    @Ignore
    internal constructor(
        position: Long,
        name: String,
        width: Int,
        height: Int,
        objects: Array<ComicPageObject>
    ) : this(0, 0, position, name, width, height, objects.asList())

    init {
        require(position >= 0L) { "Invalid position value $position" }
        require(width >= 0L) { "Invalid width value $width" }
        require(height >= 0L) { "Invalid width value $height" }
    }

    companion object {
        internal const val TABLE_NAME = "comic_book_page"

        internal const val COLUMN_ID = "id"
        internal const val COLUMN_BOOK_ID = "book_id"
        internal const val COLUMN_POSITION = "position"
        internal const val COLUMN_NAME = "name"
        internal const val COLUMN_WIDTH = "width"
        internal const val COLUMN_HEIGHT = "height"

        internal const val INDEX_BOOK_ID_POSITION = "idx_book_id_position"
    }
}